File: src/go/problems/max_depth_test.go
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

package problems

import (
	"testing"
)

func TestMaxDepth(t *testing.T) {
	var node = TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	
	result := MaxDepth(&node)
	t.Log(result)
}

func MaxDepth(root *TreeNode) int {
	return Dfs2(root, 0)
}

func Dfs2(node *TreeNode, depth int) int {
	if node == nil {
		return depth
	}

	left := Dfs2(node.Left, depth)
	right := Dfs2(node.Right, depth)

	return 1 + max1(left, right)
}

func max1(left int, right int) int {
	if left > right {
		return left
	}

	return right
}

File: src/python/problems/max_depth_test.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/
from typing import Optional
from models.tree_node import TreeNode
from core.problem_base import *

class MaxDepth(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> int:
        def Dfs(node, depth) -> int:
            if not node:
                return depth
            
            left = Dfs(node.left, depth)
            right = Dfs(node.right, depth)

            return 1 + max(left, right)

        return Dfs(root, 0)

if __name__ == '__main__':
    TestGen(MaxDepth) \
        .Add(lambda tc: tc.ParamTreeNode([3,9,20,None,None,15,7]).Result(3)) \
        .Add(lambda tc: tc.ParamTreeNode([1,None,2]).Result(2)) \
        .Run()
